Sources/BSON/DocumentIterator.swift

```diff
@@ -0,0 +1,57 @@
+import Foundation
+import NIO
+
+extension Document: Sequence {
+    // Since a `Document` is a recursive structure, we want to enforce the use of it as a subsequence of itself,
+    // instead of something like `Slice<Document>`.
+    /// The type that is returned from methods such as `dropFirst()` and `split()`.
+    public typealias SubSequence = Document
+
+    /// Returns a `Bool` indicating whether the document is empty.
+    public var isEmpty: Bool { fatalError("Unimplemented") }
+
+    /// Returns a `DocumentIterator` over the values in this `Document`.
+    public func makeIterator() -> DocumentIterator {
```

Comment:
oh, also, I had this in the design as `some IteratorProtocol`, but apparently that only works on macOS 10.15+.... dropping < 10.15 support is probably not an option for us right now (at the very least because of Evergreen not even having 10.15 to test on, lol).
so I had to make that type public again and make this return a `DocumentIterator` 😠 

opened [SWIFT-855](https://jira.mongodb.org/browse/SWIFT-855) re macOS version support policy